service/hdfs.cc

```diff
@@ -0,0 +1,177 @@
+/* hdfs.cc
+   Wolfgang Sourdeau, January 2014
+   Copyright (c) 2014 Datacratic.  All rights reserved.
+*/
+
+#include <memory>
+#include <string>
+#include <boost/iostreams/filtering_stream.hpp>
+
+#include "hdfs/hdfs.h"
+
+#include "googleurl/src/url_util.h"
+#include "jml/arch/exception.h"
+#include "jml/utils/exc_assert.h"
+#include "jml/utils/filter_streams.h"
+#include "jml/utils/guard.h"
+#include "soa/types/url.h"
+
+using namespace std;
+using namespace Datacratic;
+
+constexpr int DEFAULT_PORT(50070);
+
+namespace {
+
+struct HDFSStreamingDownloadSource {
+    HDFSStreamingDownloadSource(const string & urlStr)
+    {
+        impl.reset(new Impl(urlStr));
+    }
+
+    typedef char char_type;
+    struct category
+        : //input_seekable,
+        boost::iostreams::input,
+        boost::iostreams::device_tag,
+        boost::iostreams::closable_tag
+    { };
+
+    struct Impl {
+        Impl(const string & urlStr)
+        {
+            Url url(urlStr);
+
+            string hostname = url.host();
+            int port = url.port();
+            if (port == -1) {
+                port = DEFAULT_PORT;
+            }
+            string username = url.username();
+            string filename = url.path();
+
+            ExcAssertEqual(url.scheme(), "hdfs");
+            ExcAssert(!filename.empty());
+
+            hdfsBuilder * builder = hdfsNewBuilder();
+            string baseUrl = "hdfs://" + hostname + ":" + to_string(port);
+            hdfsBuilderSetNameNode(builder, baseUrl.c_str());
+
+            if (!username.empty()) {
+                hdfsBuilderSetUserName(builder, username.c_str());
+            }
+
+            bool ok(false);
+            ML::Call_Guard guard([&] { if (!ok) cleanup(); });
+
+            fsHandle_ = hdfsBuilderConnect(builder);
+            ExcAssert(fsHandle_ != nullptr);
+
+            if (hdfsExists(fsHandle_, filename.c_str())) {
```

Comment:
Yes, it looks like it but it's just the function name that's not very representative.